pkg/config/structure/unmarshal.go

```diff
@@ -0,0 +1,368 @@
+// does not depend upon details of the data model of the config.
+// Target struct can use of struct tag of "yaml", "json", or "mapstructure" to rename fields
+func UnmarshalKey(cfg model.Reader, key string, target interface{}) error {
+	source := newNode(reflect.ValueOf(cfg.Get(key)))
+	outValue := reflect.ValueOf(target)
+	if outValue.Kind() == reflect.Pointer {
+		outValue = reflect.Indirect(outValue)
+	}
+	if outValue.Kind() == reflect.Map {
+		return copyMap(outValue, source)
+	}
+	if outValue.Kind() == reflect.Struct {
+		return copyStruct(outValue, source)
+	}
+	if outValue.Kind() == reflect.Slice {
+		return copyList(outValue, source.AsList())
+	}
+	return fmt.Errorf("can only UnmarshalKey to struct or slice, got %v", outValue.Kind())
+}
+
+type nodeType int
+
+const (
+	invalidNodeType nodeType = iota
+	scalarNodeType
+	listNodeType
+	mapNodeType
+	structNodeType
+)
+
+type leafimpl struct {
+	val reflect.Value
+}
+
+type leaf interface {
+	GetBool() (bool, error)
+	GetInt() (int, error)
+	GetString() (string, error)
+}
+
+type nodeListImpl struct {
+	val reflect.Value
+}
+
+type nodeList interface {
+	Size() int
+	Index(int) node
+}
+
+type nodeimpl struct {
+	typ nodeType
+	val reflect.Value
+}
+
+type node interface {
+	GetChild(string) node
+	ChildrenKeys() []string
+	AsList() nodeList
+	AsScalar() leaf
+}
+
+func newNode(v reflect.Value) node {
+	if v.Kind() == reflect.Struct {
+		return &nodeimpl{typ: structNodeType, val: v}
+	}
+	if v.Kind() == reflect.Map {
+		return &nodeimpl{typ: mapNodeType, val: v}
+	}
+	if v.Kind() == reflect.Slice {
+		return &nodeimpl{typ: listNodeType, val: v}
+	}
+	if isScalar(v) {
+		return &nodeimpl{typ: scalarNodeType, val: v}
+	}
+	panic(fmt.Errorf("could not create node from: %v of type %T and kind %v", v, v, v.Kind()))
+}
+
+// GetChild returns the child node at the given key, or nil if not found
+func (n *nodeimpl) GetChild(key string) node {
+	if n.typ != mapNodeType && n.typ != structNodeType {
+		return nil
+	}
+	if n.val.Kind() == reflect.Map {
+		inner := n.val.MapIndex(reflect.ValueOf(key))
+		if !inner.IsValid() {
+			return nil
+		}
+		if inner.IsNil() {
+			return nil
+		}
+		if inner.Kind() == reflect.Interface {
+			inner = inner.Elem()
+		}
+		return newNode(inner)
+	}
+	if n.val.Kind() == reflect.Struct {
+		findex := findFieldMatch(n.val, key)
+		if findex == -1 {
+			return nil
+		}
+		inner := n.val.Field(findex)
+		if !inner.IsValid() {
+			return nil
+		}
+		if inner.Kind() == reflect.Interface {
+			inner = inner.Elem()
+		}
+		return newNode(inner)
+	}
+	return nil
+}
+
+// ChildrenKeys returns the list of keys of the children of the given node
+func (n *nodeimpl) ChildrenKeys() []string {
+	if n.val.Kind() == reflect.Map {
+		keyvals := n.val.MapKeys()
+		keys := make([]string, 0, len(keyvals))
+		for _, kv := range keyvals {
+			kinf := kv.Interface()
+			kstr, ok := kinf.(string)
+			if ok {
+				keys = append(keys, kstr)
+			}
```

Comment:
Do we want / need to support general interface types for keys? It would be nice is the design of the config could move to using strings for keys everywhere, as that simplifies the design a bit and makes it easier to understand. In this specific case, the keys are returned here and then passed to `node.GetChild` which requires a string.

Reply:
As discussed we can start like this and expand if needed.